File: 2320-find-all-k-distant-indices-in-an-array/find-all-k-distant-indices-in-an-array.rb
# @param {Integer[]} nums
# @param {Integer} key
# @param {Integer} k
# @return {Integer[]}
def find_k_distant_indices(nums, key, k)
  n = nums.length
  diff = Array.new(n + 1, 0)
  result = []

  nums.each_with_index do |num, i|
    if num == key
      start = [0, i - k].max
      ending = [n, i + k + 1].min
      diff[start] += 1
      diff[ending] -= 1 if ending < n
    end
  end

  sum = 0
  (0...n).each do |i|
    sum += diff[i]
    result << i if sum > 0
  end

  result
end

File: 2320-find-all-k-distant-indices-in-an-array/find-all-k-distant-indices-in-an-array.go
func findKDistantIndices(nums []int, key int, k int) []int {
	n := len(nums)
	diff := make([]int, n+1)
	result := []int{}

	for i := 0; i < n; i++ {
		if nums[i] == key {
			start := max(0, i-k)
			end := min(n, i+k+1)
			diff[start]++
			if end < n {
				diff[end]--
			}
		}
	}

	sum := 0
	for i := 0; i < n; i++ {
		sum += diff[i]
		if sum > 0 {
			result = append(result, i)
		}
	}

	return result
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
